### ISSUE TITLE
調査・実装: 競りQ値を最終役職確率と役職別価値へ分解して学習する

### ISSUE THREAD

**Issue by eletim** [eletim/napoleon-web#373]
Related: #202, #367, #368, #370, #372
Base branch: `feature/issue-202-non-playing-ai`

## 背景

Issue #370/#372 で、競りQ学習を `Q(s,a)` の単一terminal reward回帰として試したが、現行teacherは以下の問題を持つことが分かった。

- rewardがほぼ `-1 / 0 / 大きな正値` に集中
- suit/target間でexact tieが多い
- strongest suitとのteacher差が弱い
- repeatsを2→20へ増やしても順位は変わらず、rollout noiseが主因ではない

また、現行rewardでは action 後に最終的に

- Napoleon
- Adjutant
- Citizen
- All Pass / other

のどの役職になったかという情報を、最終Q値へ十分に構造化して使っていない。

このIssueでは、単一 `Q(s,a)` を直接学ぶ代わりに、**action後の最終役職確率と、役職ごとの価値を分解して学習する方式**を検証する。

これでもstate/action signalが弱い場合は、次段階として「最終的に何枚の絵札を取れたか」等の連続値予測へ進む。

## 基本アイデア

競り状態 `s`、競りaction `a` に対して、まず最終役職 `r` を明示的に扱う。

概念的には:

```text
P(r | s, a)
```

と

```text
V(s, a, r)
```

を学習し、最終的なaction価値を

```text
Q(s,a) = Σ_r P(r | s,a) * V(s,a,r)
```

として構成する。

ここで `r` は少なくとも:

- Napoleon
- Adjutant
- Citizen
- AllPass / NoContract / Other

を区別する。

推論時には最終役職は未知なので、**actual finalRoleを入力特徴として使ってはいけない**。finalRoleは教師ラベルとして使う。

## 目的

以下を調査・最小実装し、単一Q回帰よりもstate-conditioned signalを学びやすくなるか検証する。

1. role probability model
2. role-conditioned value model
3. それらから合成したQ値
4. 単一Q network baselineとの比較

## 1. Dataset確認

Issue #368 dataset schema v1に含まれるterminal role/result情報を確認する。

最低限、各 `(stateKey, action)` sampleについて:

- final role
- contract success/failure
- terminal reward
- candidateがauctionを取ったか
- Adjutantになったか
- Citizenになったか
- All Pass / no contractか

を再構成できることを確認する。

不足fieldがあれば、恒久dataset schema変更が必要かを判断する。

## 2. Role label設計

まず role classes を具体化する。

初版候補:

```text
Napoleon
Adjutant
Citizen
NoContract/AllPass
```

Napoleon+Adjutantの特殊ケースや、役職未確定/異常終了を別classへ分ける必要があるかコードを確認して決める。

role classはexclusiveであること。

## 3. Role probability head

`P(role | state, action)` を学習する。

推奨初版は state-action scalar network か、29-action出力をroleごとに持つmulti-head方式を比較する。

候補A:

```text
compact278 -> shared MLP -> 29 x R role logits
```

各actionについてR role probabilitiesを出す。

候補B:

```text
[state, action encoding] -> MLP -> R role logits
```

初版としてどちらが実装・評価しやすいか判断する。

教師はactual finalRole。

lossはcross entropy。

## 4. Role-conditioned value

役職になった場合の価値を別に学習する。

候補:

```text
V(s,a,r)
```

教師としてまず現行terminal rewardを使う。

ただし、役職別にreward構造がほぼ定数なら、value modelを複雑にせず、role/outcome conditional expected rewardとして扱えるか確認する。

最低限:

- Napoleonになった場合のvalue
- Adjutantになった場合のvalue
- Citizenになった場合のvalue
- NoContract時のvalue

を分けて診断する。

## 5. Q合成

role probabilityとrole-conditioned valueから:

```text
Q_role_decomposed(s,a)
```

を計算する。

validationではteacher-covered actionsのみで:

- predicted best action
- PASS vs best bid
- suit ranking
- target ranking

を評価する。

## 6. 単一Qとの比較

Issue #370のbest variantをbaselineとする。

最低限比較:

- validation MAE/RMSE
- action ranking accuracy
- best-action hit
- PASS vs best-bid accuracy
- predicted PASS率
- S/H/D/C best distribution
- strongest suit match

ただしrole probability自体についても:

- role classification accuracy
- per-role precision/recall
- confusion matrix
- calibration

を出す。

### 仮説A

`13♠` と `13♥` のterminal rewardがどちらも `-1` でも、

- Napoleonになる確率
- Citizenに戻る確率
- Adjutantになる確率

が違えば、role probability headには学習signalがある。

### 仮説B

単一reward回帰では潰れていたaction差を、role decompositionで保持できる。

### 仮説C

role decomposition後もsuit/target差がほぼ出ないなら、次に「何枚取れたか」等のcontinuous outcome predictionが必要。

この3つを明示的に検証する。

## 8. Role transition diagnostics

まず学習前にdataset自体を分析する。

state/actionごとに:

- P(Napoleon)
- P(Adjutant)
- P(Citizen)
- P(NoContract)

のempirical distributionを算出する。

特に同一stateのaction間でrole distributionがどれだけ変わるかを見る。

最低限:

- PASS vs lowest legal bidでrole probabilityがどれだけ変わるか
- suit変更でNapoleon probabilityが変わるか
- target変更でNapoleon probabilityが変わるか
- strongest suit bidでNapoleon probabilityが上がるか

ここに差が無ければnetwork実装前にその旨を明記する。

## 9. Small experiment

Issue #368 generatorで必要なdatasetを生成し、小規模学習を実施する。

推奨:

- 1000 source states程度
- candidate rotation
- action coverageはPASS + lowest legal S/H/D/C + target補完
- repeatsは1〜2でよい（#372よりnoiseはほぼゼロと分かっているため）

まずrole classifierの学習成立を見る。

その後role-conditioned Qを合成して単一Q baselineと比較する。

## 10. Success criteria

以下のどれかが確認できればrole decompositionは有望。

- role predictionがaction-index mean baselineより明確にstate-conditioned
- 同一stateでactionごとのNapoleon/Adjutant/Citizen確率差を学習できる
- 単一QよりPASS collapseが減る
- suit/target rankingが改善する
- strongest suitとの関係が改善する

逆に以下なら次段階へ進む。

- role distribution自体がaction間でほぼ同じ
- role classifierがstate/action情報を使えない
- role decompositionしてもsuit/target rankingが改善しない

その場合、次Issueでは **最終獲得絵札枚数 / Napoleon側獲得枚数 / contract margin等のcontinuous outcome予測** を教師にする。

## Tests

最低限:

1. role label extraction fixture
2. role classes exclusive
3. finalRoleをinput featureへ混入しないこと
4. role probability output shape
5. cross entropy training fixture
6. role-conditioned value loss fixture
7. Q composition fixture
8. stateKey split leakageなし
9. checkpoint save/load
10. deterministic CPU smoke
11. diagnostics correctness
12. ONNX exportする場合はPyTorch parity

## Git / PR運用

最新 `feature/issue-202-non-playing-ai` からIssue専用ブランチを作成する。

PR baseは必ず `feature/issue-202-non-playing-ai`。
main向けPRは作らない。

実装・tests・small experiment後、Codex review最大4回（review後修正最大3回）。blocking/actionable指摘がなくなればReadyにして親ブランチへmergeする。

mainにはmergeしない。

## 非目標

- runtime Q bidding agent接続
- PPO削除
- playing/adjutant/exchange再学習
- Bellman/DQN
- 獲得枚数予測の本実装（role decompositionが不十分だった場合の次Issue）

## 完了時に必ず報告

- role class定義
- role probability network構造
- role-conditioned value構造
- dataset state/sample数
- role distribution
- role prediction accuracy / confusion
- action別role probability差
- 単一Q baselineとのranking比較
- PASS率
- suit/target分布
- strongest suit match
- role decompositionが有望か否か
- 次にcontinuous獲得枚数予測へ進むべきか
- PR / merge commit / review回数
- mainへmergeしていないこと
